finagle-mysql/src/main/scala/com/twitter/finagle/mysql/ConnectionInitSql.scala

```diff
@@ -0,0 +1,30 @@
+package com.twitter.finagle.mysql
+
+import com.twitter.finagle._
+import com.twitter.finagle.mysql.param.ConnectionInitRequest
+import com.twitter.util._
+
+/**
+  * A stack module that executes a request when create a connection.
+  */
+object ConnectionInitSql {
+  val Role: Stack.Role = Stack.Role("ConnectionInitSql")
+
+  private[finagle] def module: Stackable[ServiceFactory[Request, Result]] =
+    new Stack.Module1[ConnectionInitRequest, ServiceFactory[Request, Result]] {
+      val role: Stack.Role = Role
+      val description: String = "Installs a connection init sql module in the stack"
+
+      def make(in: ConnectionInitRequest, next: ServiceFactory[Request, Result]): ServiceFactory[Request, Result] = {
+        next.flatMap { service =>
+          in.request.fold(Future.value(service)) { req =>
+            service(req).flatMap {
+              case _: OK => Future.value(service)
+              case Error(code, sqlState, message) =>
+                Future.exception(ServerError(code, sqlState, message))
+            }
+          }
+        }
+      }
```

Comment:
I think we can effectively strip out the ServiceFactory wrapper if we match on `in` before the next.flatMap call.

Reply:
changed to check `in.request` firstly.
https://github.com/twitter/finagle/pull/840/commits/c0bc82e6504441cd5cc3db80ae1d4075fccc24cc